examples/stable_diffusion_xl/gm/util/util.py

```diff
@@ -105,14 +105,23 @@ def auto_mixed_precision(network, amp_level="O0"):
     elif amp_level == "O1":
         return _auto_white_list(network, AMP_WHITE_LIST)
     elif amp_level == "O2":
-        network = _auto_black_list(
-            network,
-            AMP_BLACK_LIST
-            + [
-                nn.GroupNorm,
-            ],
-            ms.float16,
-        )
+        try:
+            _auto_black_list(
```

Comment:
I was also working on this feature in parallel with you. This entire function can be simply replaced with this:
```python
    if amp_level != "O3":
        network = amp.auto_mixed_precision(network, amp_level)
    else:
        network = amp.custom_mixed_precision(network, black_list=amp.AMP_BLACK_LIST + [nn.GroupNorm])

    if amp_level in ["O2", "O3"]:   # unwrap the network from the `_OutputTo32` wrapper
        network = network._backbone

    return network
```
And this will work on both MS2.1 and 2.2.

Reply:
I think we should request a change from the MS team to add an optional parameter for `fp32` wrapping. Sometimes, we need just `fp16` output.

Reply:
I suppose the `amp.AMP_BLACK_LIST + [nn.GroupNorm]` is the black list when amp_level equals to `O2`, not `O3`, right? Besides, I think we often want fp32 outputs to compute the loss value in full precision, because the loss function often contains softmax, log functions that prefer fp32 format.

Reply:
Blacklist is the list of layers that should be executed in full precision when amp level is `O3`.
Whitelist is the list of layers that should be executed in half precision when amp level is `O2`.

On `fp32` loss calculation better to ask @zhanghuiyao. As I understand the loss is part of the network to which AMP is applied:
https://github.com/mindspore-lab/mindone/blob/96ce7cbacc5f8ddcaa5bf293099e5f8732d82916/examples/stable_diffusion_xl/train.py#L156-L166
Therefore, the precision of the output doesn't affect the loss calculation.

Reply:
Yes, misspelled. Should be `if amp_level != "O2"`.

On `fp32` loss calculation better to ask @zhanghuiyao. As I understand the loss is part of the network to which AMP is applied:
https://github.com/mindspore-lab/mindone/blob/96ce7cbacc5f8ddcaa5bf293099e5f8732d82916/examples/stable_diffusion_xl/train.py#L156-L166
Therefore, the precision of the output doesn't affect the loss calculation in this particular case.

Reply:
The loss function calculation is included in `TrainOneStepCell`, and the `amp`  setting will take effect on the loss.